[KB-304] Stacking results (#212)
chhoumann/thesis-chemcam: report_thesis/src/sections/experiments/stacking_ensemble.tex

\subsection{Stacking Ensemble}\label{subsec:stacking_ensemble}
Given the results of the optimization process, we implemented a stacking ensemble to combine the predictions of the top-performing configurations.
As with all our experiments, we follow the procedure outlined in Section~\ref{subsec:validation_testing_procedures} to evaluate the performance of the stacking ensemble.

> For each oxide, we first identified the top-performing configurations to use in the stacking ensemble.

Reviewer: ```suggestion
For each oxide, we selected the top-performing configurations to use in the stacking ensemble.
```

> To identify the top-performing configurations to use in the stacking ensemble, we developed a category-based method, which categorizes models into distinct groups for systematic comparison.

Reviewer 2: ```suggestion
To identify the top-performing configurations for use in the stacking ensemble, we developed a category-based method that groups models into distinct categories for systematic comparison.
```

> This approach aids in selecting models that perform optimally for specific tasks, and ensures diversity in the ensemble.

Reviewer 2: ```suggestion
This approach helps in selecting models that perform optimally for specific tasks while ensuring diversity within the ensemble.
```

> We also briefly developed and employed a grid search to find the optimal ensemble of configurations, but time constraints prevented its full implementation.

Reviewer: ```suggestion
We also developed and employed a simple grid search to find the optimal ensemble of configurations, but time constraints prevented its full implementation.
```

> We also briefly developed and employed a grid search to find the optimal ensemble of configurations, but time constraints prevented its full implementation.

Reviewer 2: 'briefly developed' sounds weird imo
```suggestion
Additionally, we developed and employed a grid search to identify the optimal ensemble configurations, but time constraints prevented us from fully completing this process.
```

Finally, we employed the stacking model and evaluated its performance.
We present the results as well as the 1:1 plots in Section~\ref{subsec:stacking_ensemble_results}.

\subsubsection{Category Method \& Pipeline Selection}\label{subsec:category_method}
The category-based method organizes models into the following groups:

\begin{itemize}
    \item Gradient Boosting: \gls{gbr}, \gls{xgboost}, \gls{ngboost}
    \item Tree-Based: \gls{etr}, \gls{rf}
    \item Linear and Regularized Models: \gls{lasso}, Ridge, \gls{enet}
    \item SVM: \gls{svr}
    \item PLS: \gls{pls}
\end{itemize}

> We use the top-performing configurations that we identified in Section~\ref{sec:optimization_results} to identify configurations for the stacking ensemble.

Reviewer 2: Past tense.
```suggestion
We used the top-performing configurations identified in Section~\ref{sec:optimization_results} to select configurations for the stacking ensemble.
```

For each oxide, we filtered the trial data from the optimization experiment to create a dataset containing configuration information and performance metrics.
These datasets were sorted by \gls{rmsecv}.
We then selected unique model types for further analysis, based on the categories above.

> Our process involved selecting a set number of configurations for each category, until a maximum number of configurations for the oxide's ensemble was reached.

Reviewer 2: ```suggestion
Our process involved selecting a set number of configurations for each category until a maximum number of configurations for the oxide's ensemble was reached.
```

> To ensure diversity and limit scope, we set a maximum of 1 models per category and 3 models per stacking ensemble, with one ensemble for each oxide.

Reviewer 2: ```suggestion
To ensure diversity and limit scope, we set a maximum of one model per category and three models per stacking ensemble, with one ensemble for each oxide
```

> For each oxide, we developed pipelines to preprocess the data in accordance with the given configuration, as well as use the models with their optimal hyperparameters.

Reviewer 2: ```suggestion
For each oxide, we developed pipelines to preprocess the data according to the given configuration and utilized models with their optimal hyperparameters
```

These pipelines represent the optimal configurations for each oxide, and are used as base estimators in the stacking ensemble.

\subsubsection{Grid Search}\label{subsec:grid_search}

> To identify the optimal stacking ensemble configuration, we briefly developed and employed a grid search methodology.

Reviewer 2: ```suggestion
To identify the optimal stacking ensemble configuration, we developed and employed a grid search methodology.
```

Reviewer 2: We already said this earlier on line 8 - reference back instead?

Author: ```suggestion
```

> To identify the optimal stacking ensemble configuration, we briefly developed and employed a grid search methodology.
> This approach systematically evaluates combinations of top-performing models to determine the best ensemble based on cross-validation metrics.

Reviewer: ```suggestion
The aforementioned grid search was used to systematically evaluate combinations of top-performing models to determine the best ensemble based on cross-validation metrics.
```

> This approach systematically evaluates combinations of top-performing models to determine the best ensemble based on cross-validation metrics.

Reviewer 2: ```suggestion
This approach systematically evaluates combinations of top-performing models to determine the best ensemble configuration based on cross-validation metrics.
```

The grid search process begins by generating all possible combinations of the selected models.

> Specifically, we generate combinations of models with at least two models, and a configurable maximum number of models.

Reviewer 2: ```suggestion
Specifically, we generate combinations with at least two models, up to a configurable maximum number of models.
```

For each oxide, we constructed a pipeline for each of the top model configurations identified in Section~\ref{sec:optimization_results}.
The evaluation function iterates over each combination of base estimators, constructs a stacking ensemble pipeline, and assesses its performance using cross-validation.

The evaluation function prepares the pipeline with the current combination of base estimators and splits the data into training and testing sets using our data partitioning method.

> The stacking ensemble is then fitted on the training set.
> The meta-features generated from the base estimators are used to train the final estimator.

Reviewer 2: ```suggestion
The stacking ensemble is then fitted on the training set, and the meta-features generated from the base estimators are used to train the final estimator.
```

We compute the evaluation metrics to assess the ensemble's effectiveness.
The best combination is identified based on the lowest \gls{rmsecv} value.

> It is possible to vary the meta-learner for this process, which represents another variable to tune as part of the search process.

Reviewer 2: ```suggestion
It is possible to vary the meta-learner in this process, adding another variable to tune as part of the search process.
```

While our grid search implementation was limited to the \ce{TiO2} oxide, the methodology demonstrates the potential for identifying optimal stacking ensembles by systematically evaluating model combinations and their configurations.

> We chose not to further pursue this method due to the time constraints of this project, but believe that it is a promising approach for future work.

Reviewer: ```suggestion
We chose not to pursue this method further due to the time constraints of this project, but believe that it is a promising approach for future work.
```

Author: ```suggestion
We chose not to pursue this method further due to the time constraints of this project, but believe it is a promising approach for future work.
```

> We chose not to further pursue this method due to the time constraints of this project, but believe that it is a promising approach for future work.

Reviewer 2: ```suggestion
We chose not to further pursue this method due to the time constraints of this project, but believe it is a promising approach for future work.
```

\subsubsection{Results}\label{subsec:stacking_ensemble_results}
For each major oxide, we ran the stacking ensemble and evaluated its performance.

> We evaluated the stacking ensemble using the same configurations, varying only the meta-learner between runs.

Reviewer 2: ```suggestion
The evaluations were conducted using the same configurations, varying only the meta-learner between runs.
```

> We present the results for the \gls{enet} meta-learner, as well as the results for \gls{enet} meta-learner with \texttt{alpha} = 0.1.

Reviewer: What was the alpha of the first enet?

Author: Default

> We present the results for the \gls{enet} meta-learner, as well as the results for \gls{enet} meta-learner with \texttt{alpha} = 0.1.

Reviewer 2: ```suggestion
We present the results for the \gls{enet} meta-learner, as well as for the \gls{enet} meta-learner with \texttt{alpha} = 0.1
```

Author: ```suggestion
We present the results for the \gls{enet} meta-learner with \texttt{alpha} = 1.0, as well as for the \gls{enet} meta-learner with \texttt{alpha} = 0.1.
```

> We also present the results for \gls{svr} meta-learner, for which we used the default hyperparameters supplied by \texttt{scikit-learn}.

Reviewer 2: ```suggestion
Additionally, we present the results for the \gls{svr} meta-learner, using the default hyperparameters provided by \texttt{scikit-learn}.
```

The evaluation metrics are shown in Table~\ref{tab:stacking_ensemble_results_enet}, Table~\ref{tab:stacking_ensemble_results_enet_01}, and Table~\ref{tab:stacking_ensemble_results_svr}.

> We also provide 1:1 plots for each ensemble in Figures~\ref{fig:elasticnet_one_to_one}, \ref{fig:enetalpha01_one_to_one}, and \ref{fig:svr_one_to_one}, showing the actual versus predicted values for each oxide.

Reviewer 2: ```suggestion
Additionally, we provide 1:1 plots for each ensemble in Figures~\ref{fig:elasticnet_one_to_one}, \ref{fig:enetalpha01_one_to_one}, and \ref{fig:svr_one_to_one}, illustrating the actual versus predicted values for each oxide.
```

> A notable observation from our results is that different meta-learners exhibited varying performance levels across oxides.

Reviewer 2: ```suggestion
A notable observation from our results is that different meta-learners exhibited varying performance levels across oxides.
```

> We observed that the final predictions were strongly affected by the meta-learner, going as far as rendering some predictions nonsensical, if the wrong meta-learner was chosen.

Reviewer 2: ```suggestion
We found that the choice of meta-learner significantly impacted the final predictions, with some meta-learners leading to nonsensical results when not well-suited
```

Author: ```suggestion
We observed that the final predictions were strongly affected by the meta-learner, going as far as rendering some predictions nonsensical if the wrong meta-learner was chosen.
```

> This observation was made because we observed \ce{TiO2} predictions that were consistently near-constant values, despite varying the combination of model configurations in the \ce{TiO2} ensemble.

Reviewer 2: ```suggestion
Specifically, for \ce{TiO2}, we observed that predictions remained near-constant values despite varying the combination of model configurations in the \ce{TiO2} ensemble, highlighting the critical role of the meta-learner in the overall performance.
```

In fact, this was the reason for our implementation of the grid search process.

> The consistency we observed suggested that our hypothesis is correct: changing only the meta-learner significantly impacts the \gls{rmsecv} and prediction outcomes.

Reviewer 2: ```suggestion
The consistency of our observations supported our hypothesis: changing only the meta-learner significantly impacts the \gls{rmsecv} and prediction outcomes.
```

> We decided to investigate further, and identified potential issues with small value predictions when using a \gls{enet} meta-learner.

Reviewer 2: ```suggestion
We decided to investigate further and identified potential issues with small value predictions when using an \gls{enet} meta-learner. 
```

For example, most values for \ce{TiO2} fell between 0 and 2.5, as shown in Figure~\ref{fig:oxide_distributions}.

> The regularization term likely dominated the fitting process, leading to underfitting, which resulted in nearly constant predictions.

Reviewer 2: ```suggestion
The regularization term likely dominated the fitting process, leading to underfitting and resulting in nearly constant predictions.
```

This hypothesis was confirmed by adjusting the regularization parameter, \texttt{alpha}, in the \gls{enet}.

> Lowering \texttt{alpha} produced better outcomes, indicating that regularization adversely affected the predicted values.

Reviewer: ```suggestion
Lowering \texttt{alpha} produced better outcomes, indicating that regularization positively affected the predicted values.
```

The 1:1 plot in Figure~\ref{fig:elasticnet_one_to_one} shows the near-constant predictions for \ce{TiO2} when using a \gls{enet} meta-learner, and Figure~\ref{fig:enetalpha01_one_to_one} shows the improved predictions with \texttt{alpha} = 0.1.

> This leads us to conclude that the meta-learner's choice significantly impacts the \gls{rmsecv} and prediction outcomes.

Reviewer 2: ```suggestion
This leads us to conclude that the meta-learner's choice significantly impacts the \gls{rmsecv} and prediction outcomes.
```

The stacking approach demonstrated significant improvements in prediction accuracy as compared to the baseline we described in Section~\ref{subsec:baseline_results}, validating the efficacy of our methodology.

> We measure this by \gls{rmsep}, as this provides the fairest comparison between the baseline and stacking approach.

Reviewer 2: ```suggestion
We measured this by \gls{rmsep}, as it provides the fairest comparison between the baseline and the stacking approach.
```

> However, do also note that some evaluation metrics are worse in the stacking approach than in some individual configurations.

Reviewer 2: ```suggestion
However, it is important to note that some evaluation metrics are worse in the stacking approach than in certain individual configurations.
```

> We believe this can be improved by a significant margin with further tuning, especially of the meta-learner's hyperparameters.

Reviewer 2: ```suggestion
We believe that further tuning, particularly of the meta-learner's hyperparameters, could substantially improve these results.
```
